weepickle-implicits/src/main/scala-3/com/rallyhealth/weepickle/v1/implicits/CaseClassTo.scala

```diff
@@ -0,0 +1,197 @@
+        if (!visited(i)) {
+          defaultValues(i) match {
+            case Some(default) =>
+              elements(i) = default.apply()
+              stillMissing -= 1
+            case None =>
+              processMissing(i, visitors(i)) match {
+                case Right(extraFallbackValue) =>
+                  elements(i) = extraFallbackValue
+                  stillMissing -= 1
+                case Left(missingFieldName) =>
+                  missingKeys += missingFieldName
+              }
+          }
+        }
+        i += 1
+
+      if (!missingKeys.isEmpty) {
+        throw new Abort("missing keys in dictionary: " + missingKeys.mkString(", "))
+      }
+
+      makeWithoutDefaults(elements)
+    end makeWithDefaults
+
+    private def makeWithoutDefaults(elements: Array[AnyRef]): T =
+      fromProduct(new Product {
+        def canEqual(that: Any): Boolean = true
+        def productArity: Int = arity
+        def productElement(i: Int): Any = elements(i)
+      })
+    end makeWithoutDefaults
+
+    override def visitObject(length: Int) = new ObjVisitor[Any, T] {
+      private val elements = new Array[AnyRef](arity)
+      private val visited = elements.map(_ => false)
+      private var missing = arity // count down missing fields
+      private var currentKey: Int = outOfBounds
+
+      override def subVisitor: Visitor[_, _] =
+        if (currentKey == outOfBounds) NoOpVisitor
+        else visitors(currentKey)
+
+      override def visitKey(): Visitor[_, _] = stringVisitor
+
+      override def visitKeyValue(v: Any): Unit =
+        currentKey = fieldIndex.getOrElse(objectAttributeKeyReadMap(v.asInstanceOf[CharSequence]).toString, outOfBounds)
+
+      override def visitValue(v: Any): Unit = if (currentKey != outOfBounds) {
+        elements(currentKey) = v.asInstanceOf[AnyRef]
+        if (!visited(currentKey)) { // may get the same key more than once
+          visited(currentKey) = true
+          missing -= 1
+        }
+      }
+
+      override def visitEnd(): T =
+        if (missing == 0) makeWithoutDefaults(elements)
+        else makeWithDefaults(elements, visited, missing)
+    }
+  end CaseClassTo
+
+  inline def macroTo[T](using m: Mirror.Of[T]): To[T] = inline m match {
+    case m: Mirror.ProductOf[T] =>
+      // parallel arrays in field definition order
+      val fieldNames = macros.fieldLabels[T].map(_._1).toArray
+      val defaultValues = fieldNames.map(macros.getDefaultParams[T])
+
+      def createVisitors: Array[Visitor[_, _]] =
+        macros.summonList[Tuple.Map[m.MirroredElemTypes, To]].asInstanceOf[List[Visitor[_, _]]].toArray
+
+      val reader = new CaseClassTo[T](
+        m.fromProduct,
+        fieldNames,
+        defaultValues,
+        createVisitors
+      )
+
+      val (isSealed, discriminator) = macros.isMemberOfSealedHierarchy[T]
+      if isSealed then annotate(reader, discriminator.getOrElse("$type"), macros.fullClassName[T]._1)
```

Comment:
Do we need to use `AttributeTagged.tagName` here instead?

Reply:
Ouch! Yes, but it is in a very different place in the inheritance hierarchy. Playing around if it can be abstract in `MacrosCommon`.